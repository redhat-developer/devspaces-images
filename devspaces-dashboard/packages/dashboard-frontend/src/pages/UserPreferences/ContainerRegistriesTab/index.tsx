/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { helpers } from '@eclipse-che/common';
import {
  AlertVariant,
  Button,
  ButtonVariant,
  PageSection,
  Toolbar,
  ToolbarContent,
  ToolbarItem,
} from '@patternfly/react-core';
import { PlusCircleIcon } from '@patternfly/react-icons';
import { Table, TableBody, TableHeader } from '@patternfly/react-table';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import ProgressIndicator from '@/components/Progress';
import { lazyInject } from '@/inversify.config';
import NoRegistriesEmptyState from '@/pages/UserPreferences/ContainerRegistriesTab/EmptyState/NoRegistries';
import DeleteRegistriesModal from '@/pages/UserPreferences/ContainerRegistriesTab/Modals/DeleteRegistriesModal';
import EditRegistryModal from '@/pages/UserPreferences/ContainerRegistriesTab/Modals/EditRegistryModal';
import { AppAlerts } from '@/services/alerts/appAlerts';
import { AlertItem } from '@/services/helpers/types';
import { AppState } from '@/store';
import * as DockerConfigStore from '@/store/DockerConfig';
import { selectIsLoading, selectRegistries } from '@/store/DockerConfig/selectors';
import { RegistryEntry } from '@/store/DockerConfig/types';

type Props = MappedProps;

type State = {
  selectedItems: string[];
  registries: RegistryEntry[];
  currentRegistry: RegistryEntry;
  currentRegistryIndex: number;
  isDeleteModalOpen: boolean;
  isEditModalOpen: boolean;
};

export class ContainerRegistries extends React.PureComponent<Props, State> {
  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);

    const registries = this.props.registries;

    this.state = {
      registries,
      currentRegistry: { url: '', password: '', username: '' },
      selectedItems: [],
      currentRegistryIndex: -1,
      isEditModalOpen: false,
      isDeleteModalOpen: false,
    };
  }

  private onChangeRegistrySelection(isSelected: boolean, rowIndex: number) {
    const { registries } = this.state;
    if (rowIndex === -1) {
      const selectedItems = isSelected ? registries.map(registry => registry.url) : [];
      this.setState({ registries, selectedItems });
    } else {
      const url = registries[rowIndex].url;
      this.setState((prevState: State) => {
        return {
          registries,
          selectedItems: isSelected
            ? [...prevState.selectedItems, url]
            : prevState.selectedItems.filter(itemUrl => itemUrl !== url),
        };
      });
    }
  }

  public async componentDidMount(): Promise<void> {
    const { isLoading, requestCredentials } = this.props;
    if (!isLoading) {
      requestCredentials();
    }
  }

  public componentDidUpdate(prevProps: Props, prevState: State): void {
    if (prevProps.registries !== this.props.registries) {
      const registries = this.props.registries;
      const selectedItems: string[] = [];
      this.state.selectedItems.forEach(item => {
        if (registries.find(registry => registry.url === item) !== undefined) {
          selectedItems.push(item);
        }
      });
      this.setState({ registries, selectedItems });
    }
    if (prevState.currentRegistryIndex !== this.state.currentRegistryIndex) {
      const currentRegistry = this.getCurrentRegistry();
      this.setState({ currentRegistry });
    }
  }

  private showAlert(alert: AlertItem): void {
    this.appAlerts.showAlert(alert);
  }

  private buildRegistryRow(registry: RegistryEntry): React.ReactNode[] {
    const { url, username } = registry;

    if (/^http[s]?:\/\/.*/.test(url)) {
      return [
        <span key="host">
          <a href={url} target="_blank" rel="noreferrer">
            {url}
          </a>
        </span>,
        <span key="username">{username}</span>,
      ];
    }

    return [<span key="host">{url}</span>, <span key="username">{username}</span>];
  }

  private showOnAddRegistryModal(): void {
    this.setState({ currentRegistryIndex: -1, isEditModalOpen: true });
  }

  private showOnEditRegistryModal(rowIndex: number): void {
    this.setState({ currentRegistryIndex: rowIndex, isEditModalOpen: true });
  }

  private showOnDeleteRegistryModal(rowIndex: number): void {
    this.setState({ currentRegistryIndex: rowIndex, isDeleteModalOpen: true });
  }

  private async onDelete(registry?: RegistryEntry): Promise<void> {
    let registries: RegistryEntry[];
    if (registry === undefined) {
      registries = this.state.registries.filter(
        registry => !this.state.selectedItems.includes(registry.url),
      );
    } else {
      registries = [...this.state.registries];
      const index = registries.findIndex(_registry => _registry.url === registry.url);
      if (index > -1) {
        registries.splice(index, 1);
      } else {
        this.showAlert({
          key: 'delete-registry-fail',
          variant: AlertVariant.danger,
          title: 'Unable to find the target registry.',
        });
        return;
      }
    }
    this.setState({ isDeleteModalOpen: false, currentRegistryIndex: -1 });
    try {
      if (!registry) {
        this.setState({ selectedItems: [] });
      }
      await this.props.updateCredentials(registries);
      this.showAlert({
        key: 'delete-success',
        variant: AlertVariant.success,
        title: `Registr${registry ? 'y' : 'ies'} successfully deleted.`,
      });
    } catch (e) {
      this.showAlert({
        key: 'delete-fail',
        variant: AlertVariant.danger,
        title: `Unable to delete the registr${registry ? 'y' : 'ies'}. ${e}`,
      });
    }
  }

  private setEditModalStatus(isEditModalOpen: boolean): void {
    if (this.state.isEditModalOpen === isEditModalOpen) {
      return;
    }
    this.setState({ isEditModalOpen });
  }

  private setDeleteModalStatus(isDeleteModalOpen: boolean): void {
    if (this.state.isDeleteModalOpen === isDeleteModalOpen) {
      return;
    }
    this.setState({ isDeleteModalOpen });
  }

  private handleDelete(): void {
    this.setState({ currentRegistryIndex: -1, isDeleteModalOpen: true });
  }

  private async onRegistriesChange(registries: RegistryEntry[]): Promise<void> {
    const isEditMode = this.isEditMode;
    this.setState({ isEditModalOpen: false, currentRegistryIndex: -1 });
    try {
      await this.props.updateCredentials(registries);
      this.showAlert({
        key: 'edit-registry-success',
        variant: AlertVariant.success,
        title: `Registry successfully ${isEditMode ? 'saved' : 'added'}.`,
      });
    } catch (e) {
      this.showAlert({
        key: 'edit-registry-fail',
        variant: AlertVariant.danger,
        title: helpers.errors.getMessage(e),
      });
      try {
        await this.props.requestCredentials();
      } catch (e) {
        console.error(e);
      }
    }
  }

  private get isEditMode(): boolean {
    const { registries } = this.props;
    const { currentRegistryIndex } = this.state;
    return currentRegistryIndex > -1 && currentRegistryIndex < registries.length;
  }

  private getCurrentRegistry(): RegistryEntry {
    const { registries } = this.props;
    const { currentRegistryIndex } = this.state;
    const isEditMode = this.isEditMode;

    if (isEditMode) {
      return Object.assign({}, registries[currentRegistryIndex]);
    }
    return {
      url: '',
      password: '',
      username: '',
    };
  }

  private handleRegistryChange(editRegistry: RegistryEntry): void {
    const { registries } = this.props;
    const { currentRegistryIndex } = this.state;
    if (this.isEditMode) {
      registries[currentRegistryIndex] = editRegistry;
    } else {
      registries.push(editRegistry);
    }

    this.onRegistriesChange(registries);
  }

  render(): React.ReactNode {
    const { isLoading } = this.props;
    const {
      isEditModalOpen,
      isDeleteModalOpen,
      currentRegistry,
      currentRegistryIndex,
      selectedItems,
    } = this.state;
    const columns = ['Host', 'Username'];
    const rows =
      this.state.registries.map(registry => ({
        cells: this.buildRegistryRow(registry),
        selected: selectedItems.includes(registry.url),
      })) || [];
    const actions = [
      {
        title: 'Edit registry',
        onClick: (event, rowIndex) => {
          event.stopPropagation();
          this.showOnEditRegistryModal(rowIndex);
        },
      },
      {
        title: 'Delete registry',
        onClick: (event, rowIndex) => {
          event.stopPropagation();
          this.showOnDeleteRegistryModal(rowIndex);
        },
      },
    ];

    return (
      <React.Fragment>
        <ProgressIndicator isLoading={isLoading} />
        <EditRegistryModal
          onCancel={() => this.setEditModalStatus(false)}
          onChange={_registry => this.handleRegistryChange(_registry)}
          isEditMode={this.isEditMode}
          registry={currentRegistry}
          isOpen={isEditModalOpen}
        />
        <PageSection>
          {rows.length === 0 ? (
            <NoRegistriesEmptyState onAddRegistry={() => this.showOnAddRegistryModal()} />
          ) : (
            <React.Fragment>
              <DeleteRegistriesModal
                selectedItems={selectedItems}
                onCancel={() => this.setDeleteModalStatus(false)}
                onDelete={_registry => this.onDelete(_registry)}
                isOpen={isDeleteModalOpen}
                registry={currentRegistryIndex !== -1 ? currentRegistry : undefined}
              />
              <Toolbar id="registries-list-toolbar" className="pf-m-page-insets">
                <ToolbarContent>
                  <ToolbarItem>
                    <Button
                      variant={ButtonVariant.danger}
                      isDisabled={this.state.selectedItems.length === 0}
                      data-testid="bulk-delete-button"
                      onClick={() => this.handleDelete()}
                    >
                      Delete
                    </Button>
                  </ToolbarItem>
                  <ToolbarItem
                    alignment={{
                      md: 'alignRight',
                      lg: 'alignRight',
                      xl: 'alignRight',
                      '2xl': 'alignRight',
                    }}
                  >
                    <Button
                      variant={ButtonVariant.link}
                      data-testid="add-button"
                      icon={<PlusCircleIcon />}
                      iconPosition="left"
                      onClick={() => this.showOnAddRegistryModal()}
                    >
                      Add Container Registry
                    </Button>
                  </ToolbarItem>
                </ToolbarContent>
              </Toolbar>
              <Table
                cells={columns}
                actions={actions}
                rows={rows}
                onSelect={(event, isSelected, rowIndex) => {
                  event.stopPropagation();
                  this.onChangeRegistrySelection(isSelected, rowIndex);
                }}
                canSelectAll={true}
                aria-label="Container credentials"
                variant="compact"
              >
                <TableHeader />
                <TableBody />
              </Table>
            </React.Fragment>
          )}
        </PageSection>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  registries: selectRegistries(state),
  isLoading: selectIsLoading(state),
});

const connector = connect(mapStateToProps, DockerConfigStore.actionCreators);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(ContainerRegistries);
