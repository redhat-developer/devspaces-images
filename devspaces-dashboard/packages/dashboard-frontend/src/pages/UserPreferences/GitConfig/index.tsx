/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { helpers } from '@eclipse-che/common';
import { AlertVariant, PageSection } from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import ProgressIndicator from '@/components/Progress';
import { lazyInject } from '@/inversify.config';
import { GitConfigAddModal } from '@/pages/UserPreferences/GitConfig/AddModal';
import { GitConfigEmptyState } from '@/pages/UserPreferences/GitConfig/EmptyState';
import { GitConfigForm } from '@/pages/UserPreferences/GitConfig/Form';
import { GitConfigToolbar } from '@/pages/UserPreferences/GitConfig/Toolbar';
import { AppAlerts } from '@/services/alerts/appAlerts';
import { AppState } from '@/store';
import * as GitConfigStore from '@/store/GitConfig';
import {
  selectGitConfig,
  selectGitConfigError,
  selectGitConfigIsLoading,
} from '@/store/GitConfig/selectors';

export type Props = MappedProps;

export type State = {
  isAddEditOpen: boolean;
};

class GitConfig extends React.PureComponent<Props, State> {
  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);

    this.state = {
      isAddEditOpen: false,
    };
  }

  public async componentDidMount(): Promise<void> {
    const { gitConfigIsLoading } = this.props;

    if (gitConfigIsLoading === true) {
      this.init();
      return;
    }

    await this.handleReload();
  }

  public componentDidUpdate(prevProps: Props): void {
    this.init(prevProps);
  }

  private init(prevProps?: Props): void {
    const { gitConfigError } = this.props;
    const prevGitConfigError = prevProps?.gitConfigError;
    if (gitConfigError && gitConfigError !== prevGitConfigError) {
      this.appAlerts.showAlert({
        key: 'gitconfig-error',
        title: helpers.errors.getMessage(gitConfigError),
        variant: AlertVariant.danger,
      });
    } else if (gitConfigError === undefined && gitConfigError !== prevGitConfigError) {
      this.appAlerts.removeAlert('gitconfig-error');
    }
  }

  private async handleSave(gitConfig: GitConfigStore.GitConfig): Promise<void> {
    try {
      await this.props.updateGitConfig(gitConfig);
      this.appAlerts.showAlert({
        key: 'gitconfig-success',
        title: 'Gitconfig saved successfully.',
        variant: AlertVariant.success,
      });
    } catch (error) {
      console.error('Failed to update gitconfig', error);
    }
  }

  private async handleReload(): Promise<void> {
    try {
      await this.props.requestGitConfig();
    } catch (error) {
      console.error('Failed to reload gitconfig', error);
    }
  }

  private handleCloseAddEditModal(): void {
    this.setState({
      isAddEditOpen: false,
    });
  }

  public render(): React.ReactElement {
    const { gitConfigIsLoading, gitConfig } = this.props;
    const { isAddEditOpen } = this.state;

    const isEmpty = gitConfig === undefined;

    return (
      <React.Fragment>
        <ProgressIndicator isLoading={gitConfigIsLoading} />
        <GitConfigAddModal
          gitConfig={gitConfig}
          isOpen={isAddEditOpen}
          onCloseModal={() => this.handleCloseAddEditModal()}
          onSave={async gitConfig => {
            await this.handleSave(gitConfig);
            this.handleCloseAddEditModal();
          }}
        />
        {isEmpty ? (
          <GitConfigEmptyState />
        ) : (
          <PageSection>
            <GitConfigToolbar
              onAdd={() => {
                this.setState({
                  isAddEditOpen: true,
                });
              }}
            />
            <GitConfigForm
              gitConfig={gitConfig}
              isLoading={gitConfigIsLoading}
              onSave={async gitConfig => await this.handleSave(gitConfig)}
              onReload={async () => await this.handleReload()}
            />
          </PageSection>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  gitConfig: selectGitConfig(state),
  gitConfigIsLoading: selectGitConfigIsLoading(state),
  gitConfigError: selectGitConfigError(state),
});

const connector = connect(mapStateToProps, GitConfigStore.actionCreators, null, {
  // forwardRef is mandatory for using `@react-mock/state` in unit tests
  forwardRef: true,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(GitConfig);
