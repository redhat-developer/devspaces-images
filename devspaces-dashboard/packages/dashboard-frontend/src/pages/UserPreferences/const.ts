/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { api } from '@eclipse-che/common';

export const GIT_OAUTH_PROVIDERS: Record<api.GitOauthProvider, string> = {
  'azure-devops': 'Microsoft Azure DevOps',
  // Either Bitbucket Cloud or Bitbucket Server
  // https://github.com/eclipse-che/che-server/blob/main/wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java
  'bitbucket-server': 'Bitbucket',
  // Bitbucket Server only
  // https://github.com/eclipse-che/che-server/blob/main/wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/BitbucketServerOAuthAuthenticator.java
  bitbucket: 'Bitbucket Server (OAuth 1.0)',
  github: 'GitHub',
  github_2: 'GitHub (The second provider)',
  gitlab: 'GitLab',
} as const;

export const DEFAULT_GIT_OAUTH_PROVIDER: api.GitOauthProvider = 'github';

export const GIT_PROVIDERS: Record<api.GitProvider, string> = {
  'azure-devops': 'Microsoft Azure DevOps',
  'bitbucket-server': 'Bitbucket Server',
  github: 'GitHub',
  gitlab: 'GitLab',
} as const;

export const DEFAULT_GIT_PROVIDER: api.GitProvider = 'github';

export const GIT_PROVIDER_ENDPOINTS: Record<api.GitProvider, string> = {
  'azure-devops': 'https://dev.azure.com',
  'bitbucket-server': 'https://bitbucket.org',
  github: 'https://github.com',
  gitlab: 'https://gitlab.com',
} as const;
