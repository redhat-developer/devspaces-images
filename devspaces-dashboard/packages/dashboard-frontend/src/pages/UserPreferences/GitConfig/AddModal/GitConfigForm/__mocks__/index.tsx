/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/AddModal/GitConfigForm';
import * as GitConfigStore from '@/store/GitConfig';

export class GitConfigForm extends React.PureComponent<Props> {
  render() {
    const { onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { name: 'User One' } } as GitConfigStore.GitConfig, false)
          }
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { email: 'user-1@chetest.com', name: 'User One' } }, true)
          }
        />
      </div>
    );
  }
}
