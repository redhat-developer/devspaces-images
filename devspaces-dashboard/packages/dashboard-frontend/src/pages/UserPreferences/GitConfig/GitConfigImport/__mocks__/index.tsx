/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/GitConfigImport';

export class GitConfigImport extends React.PureComponent<Props> {
  render() {
    const { content, onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          value={content}
          onClick={() => onChange('[user]\n\tname = User One\n', false)}
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange('[user]\n\tname = User One\n\temail = user-1@chetest.com\n', true)
          }
        />
      </div>
    );
  }
}
