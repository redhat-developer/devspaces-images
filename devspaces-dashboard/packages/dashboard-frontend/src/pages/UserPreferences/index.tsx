/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { PageSection, PageSectionVariants, Tab, Tabs, Title } from '@patternfly/react-core';
import { History } from 'history';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import Head from '@/components/Head';
import ContainerRegistries from '@/pages/UserPreferences/ContainerRegistriesTab';
import GitConfig from '@/pages/UserPreferences/GitConfig';
import GitServices from '@/pages/UserPreferences/GitServices';
import PersonalAccessTokens from '@/pages/UserPreferences/PersonalAccessTokens';
import SshKeys from '@/pages/UserPreferences/SshKeys';
import { ROUTE } from '@/Routes/routes';
import { UserPreferencesTab } from '@/services/helpers/types';
import { AppState } from '@/store';
import { actionCreators } from '@/store/GitOauthConfig';
import { selectIsLoading } from '@/store/GitOauthConfig/selectors';

export type Props = {
  history: History;
} & MappedProps;

export type State = {
  activeTabKey: UserPreferencesTab;
};

export class UserPreferences extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    const activeTabKey = this.getActiveTabKey();

    this.state = {
      activeTabKey,
    };
  }

  private getActiveTabKey(): UserPreferencesTab {
    const { pathname, search } = this.props.history.location;

    if (search) {
      const searchParam = new URLSearchParams(search);
      const tab = searchParam.get('tab');
      if (
        pathname === ROUTE.USER_PREFERENCES &&
        (tab === UserPreferencesTab.CONTAINER_REGISTRIES ||
          tab === UserPreferencesTab.GITCONFIG ||
          tab === UserPreferencesTab.GIT_SERVICES ||
          tab === UserPreferencesTab.PERSONAL_ACCESS_TOKENS ||
          tab === UserPreferencesTab.SSH_KEYS)
      ) {
        return searchParam.get('tab') as UserPreferencesTab;
      }
    }

    return UserPreferencesTab.CONTAINER_REGISTRIES;
  }

  private handleTabClick(
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    activeTabKey: string | number,
  ): void {
    event.stopPropagation();
    this.props.history.push(`${ROUTE.USER_PREFERENCES}?tab=${activeTabKey}`);

    this.setState({
      activeTabKey: activeTabKey as UserPreferencesTab,
    });
  }

  render(): React.ReactNode {
    const { activeTabKey } = this.state;

    return (
      <React.Fragment>
        <Head pageName="User Preferences" />
        <PageSection variant={PageSectionVariants.light}>
          <Title headingLevel={'h1'}>User Preferences</Title>
        </PageSection>
        <Tabs
          id="user-preferences-tabs"
          style={{ backgroundColor: 'var(--pf-global--BackgroundColor--100)' }}
          activeKey={activeTabKey}
          onSelect={(event, tabKey) => this.handleTabClick(event, tabKey)}
          mountOnEnter={true}
          unmountOnExit={true}
        >
          <Tab eventKey={UserPreferencesTab.CONTAINER_REGISTRIES} title="Container Registries">
            <ContainerRegistries />
          </Tab>
          <Tab eventKey={UserPreferencesTab.GIT_SERVICES} title="Git Services">
            <GitServices />
          </Tab>
          <Tab eventKey={UserPreferencesTab.PERSONAL_ACCESS_TOKENS} title="Personal Access Tokens">
            <PersonalAccessTokens />
          </Tab>
          <Tab eventKey={UserPreferencesTab.GITCONFIG} title="Gitconfig">
            <GitConfig />
          </Tab>
          <Tab eventKey={UserPreferencesTab.SSH_KEYS} title="SSH Keys">
            <SshKeys />
          </Tab>
        </Tabs>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  isLoading: selectIsLoading(state),
});

const connector = connect(mapStateToProps, actionCreators);

type MappedProps = ConnectedProps<typeof connector>;
export default connector(UserPreferences);
