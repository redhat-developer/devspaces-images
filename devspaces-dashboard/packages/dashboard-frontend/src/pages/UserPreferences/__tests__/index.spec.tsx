/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';

import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';
import { UserPreferencesTab } from '@/services/helpers/types';
import { FakeStoreBuilder } from '@/store/__mocks__/storeBuilder';

import UserPreferences from '..';

jest.mock('../ContainerRegistriesTab');
jest.mock('../GitConfig');
jest.mock('../GitServices');
jest.mock('../PersonalAccessTokens');
jest.mock('../SshKeys');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const history = createMemoryHistory();

function getComponent(): React.ReactElement {
  const store = new FakeStoreBuilder().build();
  return (
    <Router history={history}>
      <Provider store={store}>
        <UserPreferences history={history} />
      </Provider>
    </Router>
  );
}

describe('UserPreferences', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
    snapshot.unmount();
  });

  it('should activate the Container Registries tab by default', () => {
    history.push('/user-preferences?tab=unknown-tab-name');

    renderComponent();

    expect(screen.queryByRole('tabpanel', { name: 'Container Registries' })).toBeTruthy();
  });

  describe('Location change', () => {
    it('should activate the Container Registries tab', () => {
      history.push('/user-preferences?tab=container-registries');

      renderComponent();

      expect(screen.queryByRole('tabpanel', { name: 'Container Registries' })).toBeTruthy();
    });

    it('should activate the Git Services tab', () => {
      history.push(`/user-preferences?tab=${UserPreferencesTab.GIT_SERVICES}`);

      renderComponent();

      expect(screen.queryByRole('tabpanel', { name: 'Git Services' })).toBeTruthy();
    });

    it('should activate the Personal Access Tokens tab', () => {
      history.push(`/user-preferences?tab=${UserPreferencesTab.PERSONAL_ACCESS_TOKENS}`);

      renderComponent();

      expect(screen.queryByRole('tabpanel', { name: 'Personal Access Tokens' })).toBeTruthy();
    });

    it('should activate the SSH Keys tab', () => {
      history.push(`/user-preferences?tab=${UserPreferencesTab.SSH_KEYS}`);

      renderComponent();

      expect(screen.queryByRole('tabpanel', { name: 'SSH Keys' })).toBeTruthy();
    });
  });

  describe('Tabs', () => {
    it('should activate the Container Registries tab', async () => {
      renderComponent();

      const tab = screen.getByRole('tab', { name: 'Container Registries' });
      await userEvent.click(tab);

      expect(screen.queryByRole('tabpanel', { name: 'Container Registries' })).toBeTruthy();
    });

    it('should activate the Git Services tab', async () => {
      renderComponent();

      const tab = screen.getByRole('tab', { name: 'Git Services' });
      await userEvent.click(tab);

      expect(screen.queryByRole('tabpanel', { name: 'Git Services' })).toBeTruthy();
    });

    it('should activate the Personal Access Tokens tab', async () => {
      renderComponent();

      const tab = screen.getByRole('tab', { name: 'Personal Access Tokens' });
      await userEvent.click(tab);

      expect(screen.queryByRole('tabpanel', { name: 'Personal Access Tokens' })).toBeTruthy();
    });

    it('should activate the Gitconfig tab', async () => {
      renderComponent();

      const tab = screen.getByRole('tab', { name: 'Gitconfig' });
      await userEvent.click(tab);

      expect(screen.queryByRole('tabpanel', { name: 'Gitconfig' })).toBeTruthy();
    });

    it('should activate the SSH Keys tab', async () => {
      renderComponent();

      const tab = screen.getByRole('tab', { name: 'SSH Keys' });
      await userEvent.click(tab);

      expect(screen.queryByRole('tabpanel', { name: 'SSH Keys' })).toBeTruthy();
    });
  });
});
