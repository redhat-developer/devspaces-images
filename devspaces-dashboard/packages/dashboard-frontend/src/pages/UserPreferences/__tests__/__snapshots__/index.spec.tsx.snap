// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPreferences snapshot 1`] = `
[
  <section
    className="pf-c-page__main-section pf-m-light"
  >
    <h1
      className="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe={true}
    >
      User Preferences
    </h1>
  </section>,
  <div
    className="pf-c-tabs"
    data-ouia-component-id="OUIA-Generated-Tabs-1"
    data-ouia-component-type="PF4/Tabs"
    data-ouia-safe={true}
    id="user-preferences-tabs"
    onSelect={[Function]}
    style={
      {
        "backgroundColor": "var(--pf-global--BackgroundColor--100)",
      }
    }
  >
    <ul
      className="pf-c-tabs__list"
      onScroll={[Function]}
      role="tablist"
    >
      <li
        className="pf-c-tabs__item pf-m-current"
        role="presentation"
      >
        <button
          aria-controls="pf-tab-section-ContainerRegistries-user-preferences-tabs"
          aria-selected={true}
          className="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe={true}
          id="pf-tab-ContainerRegistries-user-preferences-tabs"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Container Registries
        </button>
      </li>
      <li
        className="pf-c-tabs__item"
        role="presentation"
      >
        <button
          aria-selected={false}
          className="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe={true}
          id="pf-tab-GitServices-user-preferences-tabs"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Git Services
        </button>
      </li>
      <li
        className="pf-c-tabs__item"
        role="presentation"
      >
        <button
          aria-selected={false}
          className="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe={true}
          id="pf-tab-PersonalAccessTokens-user-preferences-tabs"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Personal Access Tokens
        </button>
      </li>
      <li
        className="pf-c-tabs__item"
        role="presentation"
      >
        <button
          aria-selected={false}
          className="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe={true}
          id="pf-tab-Gitconfig-user-preferences-tabs"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Gitconfig
        </button>
      </li>
      <li
        className="pf-c-tabs__item"
        role="presentation"
      >
        <button
          aria-selected={false}
          className="pf-c-tabs__link"
          data-ouia-component-type="PF4/TabButton"
          data-ouia-safe={true}
          id="pf-tab-SshKeys-user-preferences-tabs"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          SSH Keys
        </button>
      </li>
    </ul>
  </div>,
  <section
    aria-labelledby="pf-tab-ContainerRegistries-user-preferences-tabs"
    className="pf-c-tab-content"
    data-ouia-component-type="PF4/TabContent"
    data-ouia-safe={true}
    hidden={false}
    id="pf-tab-section-ContainerRegistries-user-preferences-tabs"
    role="tabpanel"
    tabIndex={0}
  >
    <div>
      ContainerRegistries
    </div>
  </section>,
]
`;
